engine/src/OpenGL.cpp

```diff
@@ -203,4 +216,186 @@ namespace engine::graphics {
+    void OpenGL::stencil_op(std::string sfail, std::string dfail, std::string dpass) {
+        GLenum sfail_;
+        if("keep" == sfail) sfail_ = GL_KEEP;
+        else if("zero" == sfail) sfail_ = GL_ZERO;
+        else if("replace" == sfail) sfail_ = GL_REPLACE;
+        else if("incr" == sfail) sfail_ = GL_INCR;
+        else if("invert" == sfail) sfail_ = GL_INVERT;
+        else if ("decr" == sfail) sfail_ = GL_DECR;
+
+        GLenum dfail_;
+
+        if("keep" == dfail) dfail_ = GL_KEEP;
+        else if("zero" == dfail) dfail_ = GL_ZERO;
+        else if("replace" == dfail) dfail_ = GL_REPLACE;
+        else if("incr" == dfail) dfail_ = GL_INCR;
+        else if("invert" == dfail) dfail_ = GL_INVERT;
+        else if ("decr" == dfail) dfail_ = GL_DECR;
+
+        GLenum dpass_;
+
+        if("keep" == dpass) dpass_ = GL_KEEP;
+        else if("zero" == dpass) dpass_ = GL_ZERO;
+        else if("replace" == dpass) dpass_ = GL_REPLACE;
+        else if("incr" == dpass) dpass_ = GL_INCR;
+        else if("invert" == dpass) dpass_ = GL_INVERT;
+        else if ("decr" == dpass) dpass_ = GL_DECR;
+
+        CHECKED_GL_CALL(glStencilOp, sfail_, dfail_, dpass_);
+    }
+    void OpenGL::stencil_mask(int mask) {
+        if (mask == 0)
+            CHECKED_GL_CALL(glStencilMask, 0x00);
+        else if (mask == 1)
+            CHECKED_GL_CALL(glStencilMask, 0xFF);
+    }
+
+    unsigned int OpenGL::copy_stencil_to_texture() {
+        auto platform = engine::core::Controller::get<engine::platform::PlatformController>();
+        int width     = platform->window()->width();
+        int height    = platform->window()->height();
+        unsigned int stencil_texture;
+        CHECKED_GL_CALL(glGenTextures, 1, &stencil_texture);
+        CHECKED_GL_CALL(glBindTexture, GL_TEXTURE_2D, stencil_texture);
+        CHECKED_GL_CALL(glTexImage2D, GL_TEXTURE_2D, 0, GL_R8, width, height, 0, GL_RED, GL_UNSIGNED_BYTE, nullptr);
+
+        // vec for storing buff data
+        std::vector<GLubyte> stencilData(width * height);
+
+        // read buff data
+        CHECKED_GL_CALL(glReadPixels, 0, 0, width, height, GL_STENCIL_INDEX, GL_UNSIGNED_BYTE, stencilData.data());
+
+
+        // copy to from vec to texture
+        CHECKED_GL_CALL(glBindTexture, GL_TEXTURE_2D, stencil_texture);
+        CHECKED_GL_CALL(glTexSubImage2D, GL_TEXTURE_2D, 0, 0, 0, width, height, GL_RED, GL_UNSIGNED_BYTE,
+                        stencilData.data());
+
+        CHECKED_GL_CALL(glTexParameteri, GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
+        CHECKED_GL_CALL(glTexParameteri, GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+
+        return stencil_texture;
+    }
+    void OpenGL::activate_texture(unsigned int texture, int slot) {
+        CHECKED_GL_CALL(glActiveTexture, GL_TEXTURE0 + slot);
+        CHECKED_GL_CALL(glBindTexture, GL_TEXTURE_2D, texture);
+    }
+
+    void OpenGL::bloom_color_buffers(unsigned int buffers[], int SCR_WIDTH, int SCR_HEIGHT) {
```

Comment:
U engine modulu, nazivi parametara se obično pišu malim slovima, pa bi bilo dobro uskladiti se s tom konvencijom.

Takođe, naziv funkcije i njena svrha deluju malo neusklađeno – funkcija zapravo kreira bafer, a obično funkcije koje nešto stvaraju počinju sa create, new ili make, što bi moglo učiniti namenu jasnijom.

Još jedno pitanje – da li ovo mora biti ovako usko vezano za bloom? Izgleda da se ovde kreiraju dva ista framebafera, pa možda ima smisla napraviti generičku funkciju koja kreira bafer sa color attachment-om i pozvati je dva puta umesto dupliranja koda?